src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineView.xaml.cs

```diff
@@ -0,0 +1,199 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Data;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServices.Implementation;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineView.xaml
+    /// All operations happen on the UI thread for visual studio
+    /// </summary>
+    internal sealed partial class DocumentOutlineView : UserControl, IVsCodeWindowEvents, IDisposable
+    {
+        private readonly IVsCodeWindow _codeWindow;
+        private readonly IThreadingContext _threadingContext;
+        private readonly DocumentOutlineViewModel _viewModel;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+        private readonly ComEventSink _codeWindowEventsSink;
+
+        public DocumentOutlineView(
+            DocumentOutlineViewModel viewModel,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow,
+            IThreadingContext threadingContext)
+        {
+            _codeWindow = codeWindow;
+            _threadingContext = threadingContext;
+            _viewModel = viewModel;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            DataContext = _viewModel;
+            InitializeComponent();
+            UpdateSort(SortOption.Location); // Set default sort for top-level items
+
+            // We don't think the shell is initialized lazily, so we'll Debug.Fail(), but if it was we'd still
+            // see the view created later so this will still function.
+            if (ErrorHandler.Failed(codeWindow.GetPrimaryView(out var primaryTextView)))
+                Debug.Fail("GetPrimaryView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Failed(StartTrackingView(primaryTextView)))
+                Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Succeeded(codeWindow.GetSecondaryView(out var secondaryTextView)))
+            {
+                if (ErrorHandler.Failed(StartTrackingView(secondaryTextView)))
+                    Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+            }
```

Comment:
@jasonmalinowski can you extract this out to something sensible (and shared with navbars).  This all feels highly suspect, and i imagine we'll just end up crashing in any sort of 'interesting' vs scenario.